src/modern/class/__tests__/ReactES6Class-test.js

```diff
@@ -65,6 +65,129 @@ describe('ReactES6Class', function() {
+  });
+
+  it('renders based on state using props in the constructor', function() {
+    class Foo extends React.Component {
+      constructor(props) {
+        this.state = { bar: props.initialValue };
+      }
+      changeState() {
+        this.setState({ bar: 'bar' });
+      }
+      render() {
+        if (this.state.bar === 'foo') {
+          return <div className="foo" />;
+        }
+        return <span className={this.state.bar} />;
+      }
+    }
+    var instance = test(<Foo initialValue="foo" />, 'DIV', 'foo');
+    instance.changeState();
+    test(<Foo />, 'SPAN', 'bar');
+  });
+
+  it('renders only once when setting state in componentWillMount', function() {
+    var renderCount = 0;
+    class Foo extends React.Component {
+      constructor(props) {
+        this.state = { bar: props.initialValue };
+      }
+      componentWillMount() {
+        this.setState({ bar: 'bar' });
+      }
+      render() {
+        renderCount++;
+        return <span className={this.state.bar} />;
+      }
+    }
+    test(<Foo initialValue="foo" />, 'SPAN', 'bar');
+    expect(renderCount).toBe(1);
+  });
+
+  it('should throw with non-object in the initial state property', function() {
+    [['an array'], 'a string', 1234].forEach(function(state) {
+      class Foo {
+        constructor() {
+          this.state = state;
+        }
+        render() {
+          return <span />;
+        }
+      }
+      expect(() => test(<Foo />, 'span', '')).toThrow(
+        'Invariant Violation: Foo.state: ' + 
+        'must be set to an object or null'
+      );
+    });
+  });
+
+  it('should render with null in the initial state property', function() {
+    class Foo extends React.Component {
+      constructor() {
+        this.state = null;
+      }
+      render() {
+        return <span />;
+      }
+    }
+    test(<Foo />, 'SPAN', '');
+  });
+
+  it('setState through an event handler', function() {
+    class Foo extends React.Component {
+      constructor(props) {
+        this.state = { bar: props.initialValue };
+      }
+      handleClick() {
+        this.setState({ bar: 'bar' });
+      }
+      render() {
+        return (
+          <Inner
+            name={this.state.bar}
+            onClick={this.handleClick.bind(this)}
+          />
+        );
+      }
+    }
+    test(<Foo initialValue="foo" />, 'DIV', 'foo');
+    attachedListener();
+    expect(renderedName).toBe('bar');
+  });
+
+  it('should not implicitly bind event handlers', function() {
+    class Foo extends React.Component {
+      constructor(props) {
+        this.state = { bar: props.initialValue };
+      }
+      handleClick() {
+        this.setState({ bar: 'bar' });
+      }
+      render() {
+        return (
+          <Inner
+            name={this.state.bar}
+            onClick={this.handleClick}
+          />
+        );
+      }
+    }
+    test(<Foo initialValue="foo" />, 'DIV', 'foo');
+    expect(attachedListener).toThrow()
```

Comment:
I want to reuse the same exact tests for multiple class implementations (module pattern etc). In those cases, the handler may not be on this.

Also, just because `this !== instance` doesn't mean that there isn't a wrapper around this that still allows this to be called.

Ideally it should be more precise but not quite sure how to design it in a general way.